### PR TITLE
Add events for server session state management

### PR DESCRIPTION
**Pull request by pulsastrix**
This commit adds two new events (`COAP_EVENT_SERVER_SESSION_NEW`
and `COAP_EVENT_SERVER_SESSION_DEL`), which are triggered when the
IO loop creates or deletes server-side sessions.

This can be useful for application data management: The application-defined data in `session->app` might point to a value that needs to be freed by the library user, but previously there was no way to reliably detect that a server-side session is deleted [^1] because session memory management is done inside of the IO loop.
Conversely, a library user may also want to initialize the application-specific data upon creation of a new session, which is also not reliably possible due to the fact that session creation is done inside of the IO loop upon receival of a packet from a new peer.

[^1]: While other event types like `COAP_EVENT_DTLS_CLOSED` and `COAP_EVENT_SESSION_CLOSED` already exist, these do not apply to all endpoint types:  `COAP_EVENT_DTLS_CLOSED` only applies to cleanly closed DTLS sessions, while `COAP_EVENT_SESSION_CLOSED` seems to only apply to TCP/TLS sessions.

### CODE REVIEW

File: src/coap_session.c [hunk 1]
```diff
@@ -635,11 +635,13 @@ coap_endpoint_get_session(coap_endpoint_t *endpoint,
 
   if (endpoint->context->max_idle_sessions > 0 &&
       num_idle >= endpoint->context->max_idle_sessions) {
+    coap_handle_event(oldest->context, COAP_EVENT_SERVER_SESSION_DEL, oldest);
     coap_session_free(oldest);
   }
   else if (oldest_hs) {
     coap_log(LOG_WARNING, "***%s: Incomplete session timed out\n",
              coap_session_str(oldest_hs));
+    coap_handle_event(oldest->context, COAP_EVENT_SERVER_SESSION_DEL, oldest);
```
Comment (mrdeep1) [inline src/coap_session.c:644]: `oldest `should be `oldest_hs`

File: include/coap3/coap_event.h [hunk 2]
```diff
@@ -56,7 +56,24 @@ typedef enum coap_event_t {
...
+
+/**
+ * Server session state management events
+ */
+
+ /// Called in the CoAP IO loop if a new *server-side* session is created due
+ /// to an incoming connection.
+ ///
+ /// Note that the session might not be a fully established connection yet,
+ /// it might also refer to e.g. a DTLS session in a handshaking stage.
```
Comment (mrdeep1) [inline include/coap3/coap_event.h:69]: The libcoap coding style for comments is to use `/*` ... `*/` (each new line beginning with an *), not `//` for single line comments. Please update and squash the commits into a single commit.

File: include/coap3/coap_event.h [hunk 3]
```diff
@@ -56,7 +56,24 @@ typedef enum coap_event_t {
...
+ COAP_EVENT_SERVER_SESSION_NEW = 0x4001,
+ /// Called in the CoAP IO loop if a server session is deleted (e.g. due to
+ /// inactivity or because the maximum number of idle sessions was exceeded).
+ ///
+ /// The session will still contain valid data when the event handler is
+ /// called.
```
Comment (mrdeep1) [inline include/coap3/coap_event.h:75]: As above re commenting style.

**Comment by mrdeep1** [comment]
Can you please update the man documentation in `man/coap_handler.txt.in` as well.

**Comment by pulsastrix** [comment]
Both the code comments and the manpage should now be updated accordingly.

File: man/coap_handler.txt.in [hunk 4]
```diff
@@ -207,6 +207,32 @@ COAP_EVENT_TCP_FAILED         0x1003
 COAP_EVENT_SESSION_CONNECTED  0x2001
 COAP_EVENT_SESSION_CLOSED     0x2002
 COAP_EVENT_SESSION_FAILED     0x2003
+/**
+ * (Q-)BLOCK receive errors
+ */
+COAP_EVENT_PARTIAL_BLOCK     = 0x3001,
```
Comment (obgm) [inline man/coap_handler.txt.in:213]: I suggest using the same syntax as above (i.e., no equal sign, no comma). Also for `COAP_EVENT_SERVER_SESSION_NEW` and `OAP_EVENT_SERVER_SESSION_DEL` below.

File: man/coap_handler.txt.in [hunk 5]
```diff
@@ -207,6 +207,32 @@ COAP_EVENT_TCP_FAILED         0x1003
...
+
+/**
```
Comment (obgm) [inline man/coap_handler.txt.in:215]: I think the comments should be used in the source code only. Asciidoc allows for more readable text markup outside the code environment.

**Review by obgm — COMMENTED** [review]
Thanks for doing the changes. Only some minor editorial comments from my side before merge.

File: include/coap3/coap_event.h [hunk 6]
```diff
@@ -56,7 +56,29 @@ typedef enum coap_event_t {
 /**
  * (Q-)BLOCK receive errors
  */
-  COAP_EVENT_PARTIAL_BLOCK     = 0x3001
+  COAP_EVENT_PARTIAL_BLOCK     = 0x3001,
+
+/**
```
[the remark this answers is not in the document]
Comment (pulsastrix) [reply include/coap3/coap_event.h:61]: Doxygen seems to not display subgroups for subsets of enum variants, but by adding one the first comment (in line 61) is no longer treated as part of the description for the `COAP_EVENT_SERVER_SESSION_NEW` variant, which is why I added one anyways.

[the remark this answers is not in the document]
Comment (obgm) [reply include/coap3/coap_event.h:61]: Sorry for being pedantic: Could you please use the `@`-syntax for doxygen keywords for consistency (i.e., s/\defgroup/@defgroup/)?

File: include/coap3/coap_event.h [hunk 7]
```diff
@@ -56,7 +56,32 @@ typedef enum coap_event_t {
...
+ * \defgroup srv_sess_state_events Server session state management events
+ * @{
+ */
+
+/**
+ * Called in the CoAP IO loop if a new *server-side* session is created due
+ * to an incoming connection.
+ *
+ * Note that the session might not be a fully established connection yet,
+ * it might also refer to e.g., a DTLS session in a handshaking stage.
```
Comment (obgm) [inline include/coap3/coap_event.h:71]: s/to e.g.,/to, e.g.,/
s/handshaking/handshake/

File: man/coap_handler.txt.in (hunk 5 above)
Comment (pulsastrix) [reply man/coap_handler.txt.in:214]: I removed the detailed comments for the specific event types, but kept the comments that are used to group the events  into categories for consistency reasons (as they were also included for the other categories, see lines 191, 198, 204).

**Pull request merged**